### PR TITLE
libpci: use "brute force" scan for devices

### PR DESCRIPTION
**Pull request by abrandnewusername**
To handle cases where PCI or PCIe buses are not accessible via
a bridge from bus 0, We use "brute force" scan instead of recursive
scan here. Details of these methods can be found on
https://wiki.osdev.org/PCI#.22Brute_Force.22_Scan.

Signed-off-by: Jingyao Zhou <jingyao.zhou@unsw.edu.au>

This is not the ideal solution, but a workaround for this issue: https://github.com/seL4/util_libs/issues/131

### CODE REVIEW

File: libpci/src/pci.c [hunk 1]
```diff
@@ -132,11 +132,6 @@ static void lib_pci_scan_bus(int bus);
 static void lib_pci_scan_fun(int bus, int dev, int fun)
 {
```
Comment (wom-bat) [inline libpci/src/pci.c:133]: Why not just get rid of this function, and call `libpci_add_fun()` directly from the scan?

**Review by wom-bat — CHANGES_REQUESTED** [review]

**Comment by kent-mcleod** [comment]
> This is not the ideal solution, but a workaround for this issue: #131

Would it be possible to implement this alternate scanning approach as a separate function to maintain the existing functionality for clients that already depend on it?

**Comment by wom-bat** [comment]
>  Would it be possible to implement this alternate scanning approach as a separate function to maintain the existing functionality for clients that already depend on it?
The existing functionality is just an optimisation to avoid scanning everything.   The new code brute-force scans every possible bus, and will pick up any that have been configured by the BIOS.

The existing code is wrong when:
 --- some buses are not attached via bridges but have a separate root complex
 --- the BIOS doesn't enable the bus (the new code won't work for this case either)
 --- where there is more than one PCI domain.

**Comment by kent-mcleod** [comment]
> The existing functionality is just an optimisation to avoid scanning everything. The new code brute-force scans every possible bus, and will pick up any that have been configured by the BIOS.

Is the brute force approach noticeably slower to execute?

**Comment by wom-bat** [comment]
> Is the brute force approach noticeably slower to execute?

Not that I can tell with a single PCI domain (and we don't deal with more than a single PCI domain at the moment).
Of course, to do it properly, you'd parse the DSDT and use the full 4k config space  for each device.

**Review by wom-bat — APPROVED** [review]
Looks good to me now.

**Pull request merged**